model-optimizer/extensions/middle/sparse_reshape.py

```diff
@@ -46,26 +46,23 @@ def replace_pattern(self, graph: Graph, match: dict):
         input_shape_value = sparse_reshape.in_port(1).data.get_value()
         output_shape_value = sparse_reshape.out_port(1).data.get_value()
 
+        # establish output shape if value of new shape is given as input
+        new_shape_value = sparse_reshape.in_port(2).data.get_value()
+        if output_shape_value is None and new_shape_value is not None:
+            output_shape_value = new_shape_value
+            if np.count_nonzero(output_shape_value == -1) == 1:
+                elem = np.prod(input_shape_value) // np.prod(new_shape_value[new_shape_value != -1])
+                output_shape_value[output_shape_value == -1] = elem
+
         if input_shape_value is None or output_shape_value is None:
             raise Error("Input shape and output shape values must be defined for node {}".format(sparse_reshape.id))
         if not np.array_equal(input_shape_value, output_shape_value):
             raise Error("Input shape and output shape values must be equal for node {}".format(sparse_reshape.id))
 
-        input_data_node1 = sparse_reshape.in_node(0)
-        input_data_node2 = sparse_reshape.in_node(1)
-        output_data_node1 = sparse_reshape.out_node(0)
-        output_data_node2 = sparse_reshape.out_node(1)
-        graph.remove_edge(input_data_node1.id, sparse_reshape.id)
-        graph.remove_edge(sparse_reshape.id, output_data_node1.id)
-        graph.remove_edge(input_data_node2.id, sparse_reshape.id)
-        graph.remove_edge(sparse_reshape.id, output_data_node2.id)
-        merge_data_nodes(graph, output_data_node1, input_data_node1)
-        merge_data_nodes(graph, output_data_node2, input_data_node2)
-        graph.remove_nodes_from([sparse_reshape.id, input_data_node1.id, input_data_node2.id])
-
-        # TODO: investigate why this second way does not work
-        #sparse_reshape.out_port(0).get_connection().set_source(sparse_reshape.in_port(0).get_source())
-        #sparse_reshape.out_port(1).get_connection().set_source(sparse_reshape.in_port(1).get_source())
-        #sparse_reshape.in_port(0).get_connection().set_destination(sparse_reshape.in_port(0)
-        #sparse_reshape.in_port(2).disconnect()
-        #graph.remove_nodes_from([sparse_reshape.id])
+        if 0 in sparse_reshape.out_nodes():
+            sparse_reshape.out_port(0).get_connection().set_source(sparse_reshape.in_port(0).get_source())
+
+        if 1 in sparse_reshape.out_nodes():
+            sparse_reshape.out_port(1).get_connection().set_source(sparse_reshape.in_port(1).get_source())
+
+        graph.remove_nodes_from([sparse_reshape.id])
```

Comment:
What about data node? There is no clean up after this transformation. We should care about all the extra nodes deletion or call cleanup. Here we should choose to clean up everything manualy.

Reply:
@lazarevevgeny, please advice the optimal way to have consistent graph (force_cleanup, manual cleanup or stay as is) here.

Reply:
Please, remove data node also by adding one more node to the `remove_nodes_from` function call.

Reply:
Fixed